app/lib/feed_manager.rb

```diff
@@ -131,4 +155,58 @@ def filter_from_mentions?(status, receiver_id)
 
     should_filter
   end
+
+  # Adds a status to an account's feed, returning true if a status was
+  # added, and false if it was not added to the feed. Note that this is
+  # an internal helper: callers must call trim or push updates if
+  # either action is appropriate.
+  def add_to_feed(timeline_type, account, status)
+    timeline_key = key(timeline_type, account.id)
+    reblog_key = key(timeline_type, account.id, 'reblogs')
+
+    if status.reblog?
+      # If the original status or a reblog of it is within
+      # REBLOG_FALLOFF statuses from the top, do not re-insert it into
+      # the feed
+      rank = redis.zrevrank(timeline_key, status.reblog_of_id)
+      return false if !rank.nil? && rank < FeedManager::REBLOG_FALLOFF
+
+      reblog_rank = redis.zrevrank(reblog_key, status.reblog_of_id)
+      return false unless reblog_rank.nil?
+
+      redis.zadd(timeline_key, status.id, status.id)
+      redis.zadd(reblog_key, status.id, status.reblog_of_id)
+    else
+      redis.zadd(timeline_key, status.id, status.id)
+    end
+
+    true
+  end
+
+  # Removes an individual status from a feed, correctly handling cases
+  # with reblogs, and returning true if a status was removed. As with
+  # `add_to_feed`, this does not trigger push updates, so callers must
+  # do so if appropriate.
+  def remove_from_feed(timeline_type, account, status)
+    timeline_key = key(timeline_type, account.id)
+    reblog_key = key(timeline_type, account.id, 'reblogs')
+
+    if status.reblog?
+      # 1. If the reblogging status is not in the feed, stop.
+      status_rank = redis.zrevrank(timeline_key, status.id)
+      return false if status_rank.nil?
+
+      # 2. Remove the reblogged status from the `:reblogs` zset.
+      redis.zrem(reblog_key, status.reblog_of_id)
+
+      # 3. Add the reblogged status to the feed using the reblogging
+      # status' ID as its score, and the reblogged status' ID as its
+      # value.
+      redis.zadd(timeline_key, status.id, status.reblog_of_id)
```

Comment:
I wonder why this line is needed.
On new redis timeline, it always score == member.
Thus, original status id is not vanished when reblog is added.
So, re-inserting original status id is not needed. I think.

Reply:
Updating the score will prevent from reinserting the reblogged status after the reblog has been deleted.

Reply:
To prevent re-inserting, reblog_key should be used instead of timeline_key.

Reply:
`reblog_key` is used to track existing reblogs, so you'll see removed reblogs if you do so.

Reply:
really? Feed#from_redis is reading only from timeline_key as I see and not reading from reblog_key.
and reblog_key only stores reblog data for recent REBLOG_FALLOFF items in timeline_key (FeedManager#trim).

Reply:
That is bad. It means trimming would cause reblogs of old statuses to be deleted.

Reply:
I can offer two things that will hopefully explain. The first, a long description of how these different zsets work, as I wrote it up while making the changes:

FeedManager maintains two sets of data in Redis for each feed. These feeds are keyed by `FeedManager#key`, here just `[key]`. The two sets of data we manage are:

1. `[key]`: a zset of statuses that show up in the feed
2. `[key]:reblogs`: a zset of statuses that are reblogs in the feed

When trimming the feed, we trim the feed to the last `MAX_ITEMS` statuses, and then trim the `:reblogs` zset to only include items with scores that indicate they would be in the last `REBLOG_FALLOFF` statuses.

When we add a status to a feed, we normally just add it to the feed zset (keyed by and with a value of the status ID). If the status is a reblog, instead we do the following:

1. If the status it's reblogging is in the last `REBLOG_FALLOFF` statuses on the feed, we ignore the reblog.
2. If the status it's reblogging is already reblogged in the `:reblogs` zset, we ignore the reblog, because this implies that the reblog is already in the last `REBLOG_FALLOFF` messages.
3. If we haven't given up yet, we add the reblog to the feed as usual, and also add it to the `:reblogs` zset with the value of the reblogged status' ID, and a score of the reblogging status' ID*.

*: Note that this is similar to the old behavior, where step 2 was skipped, and step 3 simply added a value with the score of the reblogging status' ID, and a value of the reblogged status's ID. We can't necessarily rely on this to be usable, because status IDs are 8-byte integers, and may not be fully representable in an IEEE double-precision float, as used by Redis for scores.

When we remove a status from a feed, we are unfortunately in the same position we had been before: because we may discard some legitimate reblogs, we cannot tell if removing the reblogged status would incorrectly remove it from our feed. Instead, if a removed status is a reblog, we do the following:

1. If the reblogging status is not in the feed, stop.
2. Remove the reblogged status from the `:reblogs` zset.
3. Add the reblogged status to the feed using the reblogging status' ID as its score, and the reblogged status' ID as its value.
4. Remove the reblogging status from the feed.

Reply:
The second thing I'll say is more handwavey but may provide a better intuition: the reblog_key zset stores data that are used to keep a status from being reblogged more than once in the general case. It is not expected to be read from in the case of a feed being read.

In the current (pre-this-PR) state of affairs, a reblog is stored with a score of the reblogging status, and a value of the reblogged status. When a reblog is deleted, that entry is removed, and one with both the score and value set to the reblogged status is added (this is in order to to avoid that status randomly disappearing if the status should have already appeared in the feed, but got a reblog attached to it first somehow. (This would happen most in the case where a user views a reblogged status and then follows the link and follows the user: their old posts should still show up in the timeline regardless of whether or not they have been reblogged.)

In this changed value, by default, we store the reblogging status' ID as both the score and value of the reblogging status. The score is stored this way to allow sorting, and yet not run into precision issues with Redis' scores being stored as double. However, as in the case now, when we remove a reblog, we can't tell if the status should still be shown or not, so we continue to err on the side of showing it. In this case, we store with a score of the old reblogging status' ID (to keep it in the same sort order), and a value of the reblogged status' ID. That's what this line is doing.

(In the future, I'd like to store a list of reasons a reblogged status should be shown to users (reblogs by which users, largely), and only re-insert a status if it has been reblogged by other users. However, that's an entirely different discussion for a different pull request.)

Reply:
I have came up a different concern in this discussion so please let me describe it here though it is a bit off-topic:
You say:
> When trimming the feed, we trim the feed to the last MAX_ITEMS statuses, and then trim the :reblogs zset to only include items with scores that indicate they would be in the last REBLOG_FALLOFF statuses.

And the code would actually work as you describe. However, the behavior sounds bad when considering `from_redis`. When a user reblogs an old status:
> we add the reblog to the feed as usual

So the reblogged, old status would get a small score. Such a status would be trimmed so quickly, which effectively makes the status invisible from `from_redis`, and `/api/v1/timelines/*`, a controller using the method.
I think such statuses should be persistent on the feed, and they are in the current implementation.

Reply:
Don't mind. I got the idea. Sorry for confusing.

Reply:
Thanks, I misunderstood this PR can solve the problem about https://github.com/tootsuite/mastodon/issues/2817#issuecomment-299501175 perfectly.
It reduces problem but not perfectly solve problem. So, 2. and 3.  are still needed. OK.